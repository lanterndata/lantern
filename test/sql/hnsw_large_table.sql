-- The goals here are:
-- - Test blockmap creation logic (triggered only after 2k vectors)
-- - Perhaps also do some lightweight benchmarking

 CREATE TABLE sift_base10k (
     id SERIAL PRIMARY KEY,
     v real[128]
);

 CREATE INDEX hnsw_idx ON sift_base10k USING hnsw (v dist_l2sq_ops) WITH (M=2, ef_construction=10, ef=4, dims=128);
 -- insert on an empty table/index
 \copy sift_base10k (v) FROM '/tmp/lanterndb/vector_datasets/siftsmall_base_arrays.csv' with csv;
SELECT V AS v4444  FROM sift_base10k WHERE id = 4444 \gset
EXPLAIN SELECT * FROM sift_base10k order by v <-> :'v4444'
LIMIT 10;
DROP INDEX hnsw_idx;
-- build index on an existing table of 10k rows
CREATE INDEX hnsw_idx ON sift_base10k USING hnsw (v dist_l2sq_ops) WITH (M=2, ef_construction=10, ef=4, dims=128);
EXPLAIN SELECT * FROM sift_base10k order by v <-> :'v4444'
LIMIT 10;
