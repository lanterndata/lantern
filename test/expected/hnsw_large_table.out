-- The goals here are:
-- - Test blockmap creation logic (triggered only after 2k vectors)
-- - Perhaps also do some lightweight benchmarking
 CREATE TABLE sift_base10k (
     id SERIAL PRIMARY KEY,
     v real[128]
);
 CREATE INDEX hnsw_idx ON sift_base10k USING hnsw (v dist_l2sq_ops) WITH (M=2, ef_construction=10, ef=4, dims=128);
INFO:  done init usearch index
INFO:  inserted 0 elements
INFO:  done saving 0 vectors
 -- insert on an empty table/index
 \copy sift_base10k (v) FROM '/tmp/lanterndb/vector_datasets/siftsmall_base_arrays.csv' with csv;
SELECT V AS v4444  FROM sift_base10k WHERE id = 4444 \gset
EXPLAIN SELECT * FROM sift_base10k order by v <-> :'v4444'
LIMIT 10;
                                                                                                                                                                                          QUERY PLAN                                                                                                                                                                                           
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.00..0.86 rows=10 width=40)
   ->  Index Scan using hnsw_idx on sift_base10k  (cost=0.00..7788.10 rows=90805 width=40)
         Order By: (v <-> '{55,61,11,4,5,2,13,24,65,49,13,9,23,37,94,38,54,11,14,14,40,31,50,44,53,4,0,0,27,17,8,34,12,10,4,4,22,52,68,53,9,2,0,0,2,116,119,64,119,2,0,0,2,30,119,119,116,5,0,8,47,9,5,60,7,7,10,23,56,50,23,5,28,68,6,18,24,65,50,9,119,75,3,0,1,8,12,85,119,11,4,6,8,9,5,74,25,11,8,20,18,12,2,21,11,90,25,32,33,15,2,9,84,67,8,4,22,31,11,33,119,30,3,6,0,0,0,26}'::real[])
(3 rows)

DROP INDEX hnsw_idx;
-- build index on an existing table of 10k rows
CREATE INDEX hnsw_idx ON sift_base10k USING hnsw (v dist_l2sq_ops) WITH (M=2, ef_construction=10, ef=4, dims=128);
INFO:  done init usearch index
INFO:  inserted 10000 elements
INFO:  done saving 10000 vectors
EXPLAIN SELECT * FROM sift_base10k order by v <-> :'v4444'
LIMIT 10;
                                                                                                                                                                                          QUERY PLAN                                                                                                                                                                                           
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.00..5.98 rows=10 width=40)
   ->  Index Scan using hnsw_idx on sift_base10k  (cost=0.00..5984.00 rows=10000 width=40)
         Order By: (v <-> '{55,61,11,4,5,2,13,24,65,49,13,9,23,37,94,38,54,11,14,14,40,31,50,44,53,4,0,0,27,17,8,34,12,10,4,4,22,52,68,53,9,2,0,0,2,116,119,64,119,2,0,0,2,30,119,119,116,5,0,8,47,9,5,60,7,7,10,23,56,50,23,5,28,68,6,18,24,65,50,9,119,75,3,0,1,8,12,85,119,11,4,6,8,9,5,74,25,11,8,20,18,12,2,21,11,90,25,32,33,15,2,9,84,67,8,4,22,31,11,33,119,30,3,6,0,0,0,26}'::real[])
(3 rows)

